docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,490 @@
+
+1. name
+2. image(ignoring the version value)
+3. function config type(`configMap` or `configPath`)
+4. relative order of the function
+
+Here is an example of the merging apply-setters function when new setters are added 
+upstream and existing setter values are updated locally. This is the most common 
+use case where kpt fails to merge them. Since `name` field is not specified, `image`
+value is used to identify and merge the function
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1
+        new-setter: new-setter-value // new setter is added
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // value of tag is updated
+
+Current Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1 // entire pipeline is overridden by upstream 
+        new-setter: new-setter-value
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // updated tag is preserved
+        new-setter: new-setter-value // new setter is pulled
+```
+
+In the above scenario, the value of the setter tag is not updated upstream, so 
+the modified local value will be preserved. But in the following example, both 
+upstream and local values change. So, similar to merging resources, upstream 
+value wins if the same fields in both upstream and local are updated.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream value changed
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-local.yaml // local value changed
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream overrides local 
+```
+
+Similarly, the upstream version wins if both upstream and local are updated, else local is preserved.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.1
+      configPath: setters.yaml
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters-updated.yaml
```

Comment:
where is `setters-updated.yaml` is coming from ? I don't see it in any of original/upstream/local ?

Reply:
It's a typo. Corrected it.